### PR TITLE
FIX-#5621: Do not preserve suboptimal partitioning on `keep_partitioning=False`

### LINKED ISSUE #5621: BUG: `.broadcast_apply_full_axis` precomputes new partition shapes incorrectly in certain cases
### Modin version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest released version of Modin.

- [X] I have confirmed this bug exists on the main branch of Modin. (In order to do this you can follow [this guide](https://modin.readthedocs.io/en/stable/getting_started/installation.html#installing-from-the-github-master-branch).)

### Reproducible Example

```python
import pandas
from modin.test.storage_formats.pandas.test_internals import (
    construct_modin_df_by_scheme,
)

import modin.config as cfg

# Has to be equal to the number of column splits
cfg.NPartitions.put(2)

md_df = construct_modin_df_by_scheme(
    pandas_df=pandas.DataFrame(
        {"a": [1, 1, 2, 2], "b": [3, 4, 5, 6], "c": [1, 2, 3, 4], "d": [4, 5, 6, 7]}
    ),
    partitioning_scheme={"row_lengths": [4], "column_widths": [2, 2]},
)

mf = md_df._query_compiler._modin_frame

print(mf._partitions.shape)  # (1, 2)
print(mf.column_widths)  # [2, 2]

res = mf.apply_full_axis(
    axis=1,
    func=lambda df: df,
    keep_partitioning=False,
    new_index=[0, 1, 2, 3],
    new_columns=["a", "b", "c", "d"],
)
print(res._partitions.shape)  # (1, 1)
print(res.column_widths)  # [4]
print(
    res._partitions[0][0].to_pandas().columns
)  # pandas.Index(["a", "b"]); the other two columns are missing!!!
```

### Issue Description

This logic of precomputing new partition lengths seems to be not always right:
https://github.com/modin-project/modin/blob/0df5fbf7a66663110878633467e4488ad37b0337/modin/core/dataframe/pandas/dataframe/dataframe.py#L2719-L2723

Here's a logic inside of `AxisPartition` that may preserve old partitioning even if we explicitly passed `keep_partitioning=False`:
https://github.com/modin-project/modin/blob/0df5fbf7a66663110878633467e4488ad37b0337/modin/core/dataframe/pandas/partitioning/axis_partition.py#L196-L203
If we hit the case then we result in an inconsistency of `new_column_widths` and the actual partitioning. In the reproducer above, the result of `.apply_full_axis` has `column_widths == [4]` however actual widths were unchanged (it's still `[2, 2]`). This causes the second column partition to be filtered out as judging by the `columns_widths == [4]` it's considered to be empty.

There's a parameter for `AxisPartition.apply` called `maintain_partitioning` that controls this behavior, if it's `False` then the old partitioning is never preserved, the only problem is that it's always `True` besides of `keep_partitioning` parameter. 
https://github.com/modin-project/modin/blob/0df5fbf7a66663110878633467e4488ad37b0337/modin/core/dataframe/pandas/partitioning/axis_partition.py#L53

It seems that all we have to do here is just forward `keep_partitioning` parameter to `AxisPartition.apply` as `maintain_partitioning=keep_partitioning`.

### Expected Behavior

...

### Error Logs

<details>

```python-traceback

Replace this line with the error backtrace (if applicable).

```

</details>

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

This may increase data transfer between workers when `keep_partitioning=False`, however, don't we always want the data to be shuffled when this parameter is passed?

The problem is that `keep_partitioning=False` is the default one for `PartitionManager.map_axis_partition`, so it may affect cases where some code calls this method without explicitly specifying this argument:
https://github.com/modin-project/modin/blob/0df5fbf7a66663110878633467e4488ad37b0337/modin/core/dataframe/pandas/partitioning/partition_manager.py#L537
Most of the cases we explicitly specify `keep_partitioning=True`, there's only one case that use the default value for this parameter:
https://github.com/modin-project/modin/blob/0df5fbf7a66663110878633467e4488ad37b0337/modin/core/dataframe/pandas/dataframe/dataframe.py#L2812-L2816

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #5621 
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 1]
```diff
@@ -1126,8 +1133,15 @@ def _reorder_labels(self, row_positions=None, col_positions=None):
...
+            # We want to preserve the frame's partitioning so passing in ``keep_partitioning=True``
+            # in order to use the cached `column_widths` values for the new frame.
+            # If the frame's is re-partitioned using the "standard" partitioning,
+            # then knowing that, we can compute new column widths.
             ordered_cols = self._partition_mgr_cls.map_axis_partitions(
-                1, ordered_rows, lambda df: df.iloc[:, col_positions]
+                1,
+                ordered_rows,
+                lambda df: df.iloc[:, col_positions],
+                keep_partitioning=True,
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:1144]: so we can modify the condition below as the number of partitions will now always be the same?
```diff
-            if self._partitions.shape[1] != ordered_cols.shape[1] or len(
-                col_idx
-            ) != len(self.columns):
+            if len(col_idx) != len(self.columns):
```

Comment (YarShev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1142]: yes, removed

**Review by dchigarev — COMMENTED** [review]
LGTM! @YarShev thanks for fixing problems with the initial approach!

(can't actually approve it though as the PR was created under my name)

**Review by YarShev — APPROVED** [review]

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 2]
```diff
@@ -1126,16 +1131,21 @@ def _reorder_labels(self, row_positions=None, col_positions=None):
...
             )
             col_idx = self.columns[col_positions]
             if new_dtypes is not None:
                 new_dtypes = self._dtypes.iloc[col_positions]
 
-            if self._partitions.shape[1] != ordered_cols.shape[1] or len(
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:1136]: Now there can't be a case where `self._partitions.shape[1] != ordered_cols.shape[1]` and `len(col_idx) == len(self.columns)`?

Comment (YarShev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1136]: No, because we pass `keep_partitioning=True` in.

**Review by anmyachev — APPROVED** [review]

**Pull request merged**